piquasso/api/branch.py

```diff
@@ -0,0 +1,102 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Tuple, Union
+
+from .state import State
+
+from fractions import Fraction
+
+
+class Branch:
+    r"""Represents a branching in the computation caused by a measurement.
+
+    Describes a branch in the computation tree that arises due to a measurement
+    operation. Each branch contains the post-measurement quantum state, the outcomes
+    of the measurements that led to this branch, and the frequency (or weight) of
+    this branch in the overall ensemble. The frequency indicates how often this
+    particular outcome occurred relative to the total number of shots (simulations)
+    performed.
+
+    The data in the branch is illustrated as follows. Consider the following setup,
+    where a measurement is performed with two possible outcomes, 0 and 1:
+
+    .. raw:: html
+        :class: only-light
+
+        <div style="text-align: center;">
+            <img
+                src="../_static/branch_illustration.svg"
+                style="width: 35%; max-width: 500px; height: auto;"
+                alt="Branch illustration"
+            >
+        </div>
+
+    .. raw:: html
+        :class: only-dark
+
+        <div style="text-align: center;">
+            <img
+                src="../_static/branch_illustration_dark.svg"
+                style="width: 35%; max-width: 500px; height: auto;"
+                alt="Branch illustration"
+            >
+        </div>
+
+    In this case, two branches are created, one for each possible outcome of the
+    measurement. The branches contain the post-measurement quantum states
+    :math:`|\psi_0\rangle` and :math:`|\psi_1\rangle`, the outcomes 0 and 1, and
+    their corresponding probabilities :math:`p_0` and :math:`p_1`,
+    corresponding to each outcome. Since the ciruit is executed `shots` times, instead
+    of the probabilities, the frequencies are stored in the branches, which are
+    calculated as :math:`f_i = n_i / \text{shots}`, where :math:`n_i` is the number of
+    times the outcome :math:`i` was observed.
+
+    :ivar state:
+        The post-measurement quantum state corresponding to the branch. If the branch
+        has no associated quantum state (e.g., all the modes are measured), then
+        this is set to `None`.
+    :ivar outcome:
+        The accumulated outcome corresponding to the measurement(s) that caused the
+        branching. If no measurements were performed, then this is is set to `None`.
+    :ivar frequency:
+        The contribution of the state in the final ensemble. If the branch is the result
+        of a measurement with multiple possible outcomes, then this is set to the
+        ratio of the number of times `outcome` appears in the simulation, divided by
+        `shots` (which is the total number of samples taken). If the branch is not the
```

Comment:
I think we discussed during our conversations that the denominator of frequency is not always all the shots - has this changed, or am I misremembering something?